src/lib/statistics/cardinality_estimator.hpp

```diff
@@ -61,17 +71,42 @@ class CardinalityEstimator {
    * @{
    *
    * For increased cardinality estimation performance:
-   * Promises to this CardinalityEstimator that it will only be used to estimate Cardinalities of plans that consist
-   * of the Vertices and Predicates in @param JoinGraph. This enables using the JoinGraphStatisticsCache during
-   * Cardinality estimation.
+   * Promises to this CardinalityEstimator that it will only be used to estimate cardinalities of plans that consist
+   * of the vertices and predicates in @param JoinGraph. This enables using the JoinGraphStatisticsCache during
+   * cardinality estimation.
    */
   void guarantee_join_graph(const JoinGraph& join_graph);
 
   /**
    * For increased cardinality estimation performance:
-   * Promises to this CardinalityEstimator that it will only be used to estimate bottom-up
-   * constructed plans. That is, the Cost/Cardinality of a node, once constructed, never changes.
-   * This enables the usage of a <lqp-ptr> -> <cost> cache.
+   * Promises to this CardinalityEstimator that it will only be used to estimate bottom-up constructed plans. Thus, the
+   * cardinalities/statistics of nodes, once constructed, never change. This enables the usage of an
+   * <lqp-ptr> -> <statistics> cache.
+   *
+   * Image the following simple example of predicate reordering. Let's say we have a table R with 100'000 tuples, a
+   * PredicateNode A with a selectivity of 0.3, and a PredicateNode B with a selectivity of 0.5. There are also more
+   * nodes on top of the subplan. The inital structure (ordered by appearence in the SQL query) looks like that:
+   *
+   *             ...
+   *              |
+   *              | 15'000 (30'000 * 0.5)
+   *              |
+   *        PredicateNode B
+   *              |
+   *              | 30'000 (100'000 * 0.3)
+   *              |
+   *        PredicateNode A
+   *              |
+   *              | 100'000
+   *              |
+   *       StoredTableNode R
+   *
+   * First, we go top-down. As we already estimated nodes further up in the LQP, the statistics/cardinalities of the
+   * PredicateNodes have been cached. PredicateNode B yields a smaller cardinality even though it has a worse
+   * selectivity, and we would falsely move it below PredicateNode A.
```

Comment:
I don't get it. :D
How does a wrong ordering relate to the cardinality estimation here? It's not clear how caching has an impact here.

Reply:
To get estimate the statistics of a node, we must know the statistics of input nodes, so we recursively estimate them. With caching, we store all estimated statistics. That's it already. When going top-down, ALL statistics will be saved once we estimate the root node. When we want to reorder the predicates, we always get the statistics that were estimated in the original order (and not, as we would like to, directly on the node below the predicate chain).

Reply:
```c++
   * First, we go top-down. As we already estimated nodes further up in the LQP, the statistics/cardinalities of the
   * PredicateNodes have been cached and will stay unchanged whenever we try to re-estimate them. When placed directly
   * above the StoredTableNode R, PredicateNode B yields a smaller cardinality than PredicateNode A even though it has a
   * worse selectivity, and we would falsely move it below PredicateNode A.
   *
   * Second, we go bottom-up. No statistics have been cached yet, we correctly estimate the cardinality based on the
   * selectivity, and we preserve the predicate order before continuing to the nodes further up. Note that during
   * predicate reordering, we do not cache the statistics of the PredicateNodes when deciding on their placement, but
   * only when we estimate nodes above.
```
better?

Reply:
Let's talk in person.